glib-macros/src/variant_derive.rs

```diff
@@ -202,3 +205,360 @@ pub fn derive_variant_for_struct(
 
     derived.into()
 }
+
+enum EnumMode {
+    String,
+    Repr(Ident),
+    Enum { repr: bool },
+    Flags { repr: bool }
+}
+
+impl EnumMode {
+    fn tag_type(&self) -> char {
+        match self {
+            EnumMode::String => 's',
+            EnumMode::Repr(repr) => match repr.to_string().as_str() {
+                "i8" | "i16" => 'n',
+                "i32" => 'i',
+                "i64" => 'x',
+                "u8" => 'y',
+                "u16" => 'q',
+                "u32" => 'u',
+                "u64" => 't',
+                _ => unimplemented!(),
+            },
+            EnumMode::Enum { repr } => if *repr { 'i' } else { 's' },
+            EnumMode::Flags { repr } => if *repr { 'u' } else { 's' },
+        }
+    }
+}
+
+fn derive_variant_for_enum(
+    ident: Ident,
+    generics: Generics,
+    data_enum: syn::DataEnum,
+    mode: EnumMode,
+) -> TokenStream {
+    let glib = crate_ident_new();
+
+    let has_data = data_enum
+                .variants
+                .iter()
+                .any(|v| !matches!(v.fields, syn::Fields::Unit));
+
+    let tag_type = mode.tag_type();
+
+    let static_variant_type = if has_data {
+        format!("({}v)", tag_type)
+    } else {
+        tag_type.to_string()
+    };
+
+    let (to_variant, from_variant) = if has_data {
```

Comment:
Maybe move the two branches of this `if` into two separate functions. That's easier to read.